File: backend/app/models/membership.py
from sqlalchemy import Column, String, Numeric, Integer, JSON, Boolean, Date, ForeignKey, Enum as SQLEnum
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.base import Base, BaseModel
import enum


class DurationType(str, enum.Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    ANNUAL = "annual"
    LIFETIME = "lifetime"


class MembershipStatus(str, enum.Enum):
    ACTIVE = "active"
    FROZEN = "frozen"
    EXPIRED = "expired"
    CANCELLED = "cancelled"


class MembershipPlan(Base, BaseModel):
    __tablename__ = "membership_plans"

    organization_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    name = Column(String(100), nullable=False)
    description = Column(String(1000), nullable=True)
    price = Column(Numeric(10, 2), nullable=False)
    duration_days = Column(Integer, nullable=False)
    duration_type = Column(SQLEnum(DurationType), nullable=False)
    setup_fee = Column(Numeric(10, 2), default=0)
    features = Column(JSON, default=dict)
    access_hours = Column(JSON, default=dict)
    is_active = Column(Boolean, default=True)

    # Relationships
    organization = relationship("Organization", back_populates="membership_plans")
    memberships = relationship("Membership", back_populates="plan")


class Membership(Base, BaseModel):
    __tablename__ = "memberships"

    organization_id = Column(UUID(as_uuid=True), ForeignKey("organizations.id"), nullable=False, index=True)
    member_id = Column(UUID(as_uuid=True), ForeignKey("members.id"), nullable=False, index=True)
    plan_id = Column(UUID(as_uuid=True), ForeignKey("membership_plans.id"), nullable=False)
    start_date = Column(Date, nullable=False)
    end_date = Column(Date, nullable=False)
    auto_renew = Column(Boolean, default=True)
    status = Column(SQLEnum(MembershipStatus), default=MembershipStatus.ACTIVE)
    freeze_start_date = Column(Date, nullable=True)
    freeze_end_date = Column(Date, nullable=True)
    cancellation_date = Column(Date, nullable=True)
    cancellation_reason = Column(String(500), nullable=True)

    # Relationships
    member = relationship("Member", back_populates="memberships")
    plan = relationship("MembershipPlan", back_populates="memberships")
    payments = relationship("Payment", back_populates="membership")

File: mobile-app/lib/models/membership.dart
class Membership {
  final String id;
  final String memberId;
  final String planId;
  final String planName;
  final DateTime startDate;
  final DateTime endDate;
  final bool autoRenew;
  final MembershipStatus status;
  final DateTime? freezeStartDate;
  final DateTime? freezeEndDate;
  final String? cancellationReason;

  Membership({
    required this.id,
    required this.memberId,
    required this.planId,
    required this.planName,
    required this.startDate,
    required this.endDate,
    required this.autoRenew,
    required this.status,
    this.freezeStartDate,
    this.freezeEndDate,
    this.cancellationReason,
  });

  factory Membership.fromJson(Map<String, dynamic> json) {
    return Membership(
      id: json['id'],
      memberId: json['member_id'],
      planId: json['plan_id'],
      planName: json['plan_name'] ?? '',
      startDate: DateTime.parse(json['start_date']),
      endDate: DateTime.parse(json['end_date']),
      autoRenew: json['auto_renew'] ?? false,
      status: MembershipStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => MembershipStatus.active,
      ),
      freezeStartDate: json['freeze_start_date'] != null
          ? DateTime.parse(json['freeze_start_date'])
          : null,
      freezeEndDate: json['freeze_end_date'] != null
          ? DateTime.parse(json['freeze_end_date'])
          : null,
      cancellationReason: json['cancellation_reason'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'member_id': memberId,
      'plan_id': planId,
      'plan_name': planName,
      'start_date': startDate.toIso8601String(),
      'end_date': endDate.toIso8601String(),
      'auto_renew': autoRenew,
      'status': status.name,
      'freeze_start_date': freezeStartDate?.toIso8601String(),
      'freeze_end_date': freezeEndDate?.toIso8601String(),
      'cancellation_reason': cancellationReason,
    };
  }

  int get daysRemaining {
    final now = DateTime.now();
    if (now.isAfter(endDate)) return 0;
    return endDate.difference(now).inDays;
  }

  bool get isExpired => DateTime.now().isAfter(endDate);
  bool get isActive => status == MembershipStatus.active && !isExpired;
}

enum MembershipStatus {
  active,
  frozen,
  expired,
  cancelled,
}

class MembershipPlan {
  final String id;
  final String name;
  final String description;
  final double price;
  final int durationDays;
  final String durationType;
  final double setupFee;
  final List<String> features;
  final Map<String, dynamic>? accessHours;
  final bool isActive;

  MembershipPlan({
    required this.id,
    required this.name,
    required this.description,
    required this.price,
    required this.durationDays,
    required this.durationType,
    required this.setupFee,
    required this.features,
    this.accessHours,
    required this.isActive,
  });

  factory MembershipPlan.fromJson(Map<String, dynamic> json) {
    return MembershipPlan(
      id: json['id'],
      name: json['name'],
      description: json['description'] ?? '',
      price: (json['price'] ?? 0).toDouble(),
      durationDays: json['duration_days'] ?? 30,
      durationType: json['duration_type'] ?? 'monthly',
      setupFee: (json['setup_fee'] ?? 0).toDouble(),
      features: List<String>.from(json['features'] ?? []),
      accessHours: json['access_hours'],
      isActive: json['is_active'] ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'price': price,
      'duration_days': durationDays,
      'duration_type': durationType,
      'setup_fee': setupFee,
      'features': features,
      'access_hours': accessHours,
      'is_active': isActive,
    };
  }
}
